torch/_subclasses/functional_tensor.py

```diff
@@ -190,23 +191,32 @@ def mark_mutation_hidden_from_autograd(self) -> None:
 
 
 class FunctionalTensorMode(TorchDispatchMode):
-    def __init__(self):
+    def __init__(self, pre_dispatch=False):
         self.is_on_stack = False
         self.enter_stack = []
         # Indicates to our torch_dispatch dispatching infra that
         # this is an "infra" mode with lower dispatching precedence.
         self._mode_key = torch._C._TorchDispatchModeKey.FUNCTIONAL
         # This will be turned off later for pre-dispatch functionalization
         self.decompose_composite_implicit_ops = True
+        self.pre_dispatch = pre_dispatch
+        self._dispatch_key = torch._C.DispatchKey.PreDispatch if pre_dispatch else None  # type: ignore[attr-defined]
 
     # No-op if FunctionalTensorMode is already in use
     def __enter__(self):
-        if (
-            torch._C._get_dispatch_mode(torch._C._TorchDispatchModeKey.FUNCTIONAL)
-            is None
-        ):
-            self.enter_stack.append(True)
+        def _get_prev_mode():
```

Comment:
Same comment as https://github.com/pytorch/pytorch/pull/113728/files#r1427430598 - if we had that util then we could just use it here (and it would give us the nice invariant that you can never had a functional mode in two stacks at once)